syntax = "proto3";

option csharp_namespace = "Gateway.Protos.Agent";

package agent;

service AgentManagementService {
  rpc RegisterAgent (RegisterAgentRequest) returns (RegisterAgentResponse);
  rpc UpdateAgentStatus (UpdateAgentStatusRequest) returns (UpdateAgentStatusResponse);
  rpc GetAgent (GetAgentRequest) returns (GetAgentResponse);
  rpc GetAgentsByComputer (GetAgentsByComputerRequest) returns (GetAgentsByComputerResponse);
  rpc GetAllAgents (GetAllAgentsRequest) returns (GetAllAgentsResponse);
  rpc DeleteAgent (DeleteAgentRequest) returns (DeleteAgentResponse);

  rpc GetAgentPolicy (GetAgentPolicyRequest) returns (GetAgentPolicyResponse);
  rpc UpsertAgentPolicy (UpsertAgentPolicyRequest) returns (UpsertAgentPolicyResponse);
  rpc DeleteAgentPolicy (DeleteAgentPolicyRequest) returns (DeleteAgentPolicyResponse);
  rpc GetAgentPolicyVersions (GetAgentPolicyVersionsRequest) returns (GetAgentPolicyVersionsResponse);
  rpc RestoreAgentPolicyVersion (RestoreAgentPolicyVersionRequest) returns (RestoreAgentPolicyVersionResponse);
  rpc GetPendingAgentCommands (GetPendingAgentCommandsRequest) returns (GetPendingAgentCommandsResponse);
  rpc GetAgentCommands (GetAgentCommandsRequest) returns (GetAgentCommandsResponse);
  rpc CreateAgentCommand (CreateAgentCommandRequest) returns (CreateAgentCommandResponse);
  rpc AckAgentCommand (AckAgentCommandRequest) returns (AckAgentCommandResponse);
  
  rpc CreateSyncBatch (CreateSyncBatchRequest) returns (CreateSyncBatchResponse);
  rpc UpdateSyncBatch (UpdateSyncBatchRequest) returns (UpdateSyncBatchResponse);
  rpc GetSyncBatch (GetSyncBatchRequest) returns (GetSyncBatchResponse);
  rpc GetSyncBatchesByAgent (GetSyncBatchesByAgentRequest) returns (GetSyncBatchesByAgentResponse);
  rpc GetPendingSyncBatches (GetPendingSyncBatchesRequest) returns (GetPendingSyncBatchesResponse);
}

message RegisterAgentRequest {
  int64 computer_id = 1;
  string version = 2;
  string config_version = 3;
}

message RegisterAgentResponse {
  bool success = 1;
  string message = 2;
  Agent agent = 3;
}

message UpdateAgentStatusRequest {
  int64 agent_id = 1;
  string status = 2;
  string config_version = 3;
}

message UpdateAgentStatusResponse {
  bool success = 1;
  string message = 2;
  Agent agent = 3;
}

message GetAgentRequest {
  int64 agent_id = 1;
}

message GetAgentResponse {
  bool success = 1;
  string message = 2;
  Agent agent = 3;
}

message GetAgentsByComputerRequest {
  int64 computer_id = 1;
}

message GetAgentsByComputerResponse {
  bool success = 1;
  string message = 2;
  repeated Agent agents = 3;
}

message GetAllAgentsRequest {
  string status = 1;
  int32 page = 2;
  int32 page_size = 3;
}

message GetAllAgentsResponse {
  bool success = 1;
  string message = 2;
  repeated Agent agents = 3;
  int32 total_count = 4;
}

message DeleteAgentRequest {
  int64 agent_id = 1;
}

message DeleteAgentResponse {
  bool success = 1;
  string message = 2;
}

message GetAgentPolicyRequest {
  int64 agent_id = 1;
}

message GetAgentPolicyResponse {
  bool success = 1;
  string message = 2;
  AgentPolicy policy = 3;
}

message UpsertAgentPolicyRequest {
  AgentPolicy policy = 1;
}

message UpsertAgentPolicyResponse {
  bool success = 1;
  string message = 2;
  AgentPolicy policy = 3;
}

message DeleteAgentPolicyRequest {
  int64 agent_id = 1;
}

message DeleteAgentPolicyResponse {
  bool success = 1;
  string message = 2;
}

message GetAgentPolicyVersionsRequest {
  int64 agent_id = 1;
  int32 page = 2;
  int32 page_size = 3;
}

message GetAgentPolicyVersionsResponse {
  bool success = 1;
  string message = 2;
  repeated AgentPolicyVersion versions = 3;
  int32 total_count = 4;
}

message RestoreAgentPolicyVersionRequest {
  int64 agent_id = 1;
  int64 version_id = 2;
  string requested_by = 3;
}

message RestoreAgentPolicyVersionResponse {
  bool success = 1;
  string message = 2;
  AgentPolicy policy = 3;
  AgentPolicyVersion restored_from = 4;
}

message GetPendingAgentCommandsRequest {
  int64 agent_id = 1;
  int32 limit = 2;
}

message GetPendingAgentCommandsResponse {
  bool success = 1;
  string message = 2;
  repeated AgentCommand commands = 3;
}

message GetAgentCommandsRequest {
  int64 agent_id = 1;
  string status = 2;
  int32 page = 3;
  int32 page_size = 4;
}

message GetAgentCommandsResponse {
  bool success = 1;
  string message = 2;
  repeated AgentCommand commands = 3;
  int32 total_count = 4;
}

message CreateAgentCommandRequest {
  int64 agent_id = 1;
  string type = 2;
  string payload_json = 3;
  string requested_by = 4;
}

message CreateAgentCommandResponse {
  bool success = 1;
  string message = 2;
  AgentCommand command = 3;
}

message AckAgentCommandRequest {
  int64 command_id = 1;
  string status = 2;
  string result_message = 3;
}

message AckAgentCommandResponse {
  bool success = 1;
  string message = 2;
  AgentCommand command = 3;
}

message CreateSyncBatchRequest {
  int64 agent_id = 1;
  string batch_id = 2;
  int32 records_count = 3;
}

message CreateSyncBatchResponse {
  bool success = 1;
  string message = 2;
  SyncBatch batch = 3;
}

message UpdateSyncBatchRequest {
  int64 batch_id = 1;
  string status = 2;
  int32 records_count = 3;
}

message UpdateSyncBatchResponse {
  bool success = 1;
  string message = 2;
  SyncBatch batch = 3;
}

message GetSyncBatchRequest {
  int64 batch_id = 1;
}

message GetSyncBatchResponse {
  bool success = 1;
  string message = 2;
  SyncBatch batch = 3;
}

message GetSyncBatchesByAgentRequest {
  int64 agent_id = 1;
  string status = 2;
  int32 page = 3;
  int32 page_size = 4;
}

message GetSyncBatchesByAgentResponse {
  bool success = 1;
  string message = 2;
  repeated SyncBatch batches = 3;
  int32 total_count = 4;
}

message GetPendingSyncBatchesRequest {
  int32 page = 1;
  int32 page_size = 2;
}

message GetPendingSyncBatchesResponse {
  bool success = 1;
  string message = 2;
  repeated SyncBatch batches = 3;
  int32 total_count = 4;
}

message Agent {
  int64 id = 1;
  int64 computer_id = 2;
  string version = 3;
  string status = 4;
  string last_heartbeat = 5;
  string config_version = 6;
  string offline_since = 7;
}

message SyncBatch {
  int64 id = 1;
  int64 agent_id = 2;
  string batch_id = 3;
  string status = 4;
  string synced_at = 5;
  int32 records_count = 6;
}

message AgentPolicy {
  int64 id = 1;
  int64 agent_id = 2;
  int64 computer_id = 3;
  string policy_version = 4;
  int32 collection_interval_sec = 5;
  int32 heartbeat_interval_sec = 6;
  int32 flush_interval_sec = 7;
  bool enable_process_collection = 8;
  bool enable_browser_collection = 9;
  bool enable_active_window_collection = 10;
  bool enable_idle_collection = 11;
  int32 idle_threshold_sec = 12;
  int32 browser_poll_interval_sec = 13;
  int32 process_snapshot_limit = 14;
  float high_risk_threshold = 15;
  bool auto_lock_enabled = 16;
  bool admin_blocked = 17;
  string blocked_reason = 18;
  repeated string browsers = 19;
  string updated_at = 20;
  string signature = 21;
  string signature_key_id = 22;
  string signature_alg = 23;
}

message AgentPolicyVersion {
  int64 id = 1;
  int64 agent_id = 2;
  string policy_version = 3;
  string change_type = 4;
  string changed_by = 5;
  string created_at = 6;
  string snapshot_json = 7;
}

message AgentCommand {
  int64 id = 1;
  int64 agent_id = 2;
  string type = 3;
  string payload_json = 4;
  string status = 5;
  string requested_by = 6;
  string result_message = 7;
  string created_at = 8;
  string acknowledged_at = 9;
  string signature = 10;
  string signature_key_id = 11;
  string signature_alg = 12;
}
